include/irr/asset/IAsset.h

```diff
@@ -173,14 +173,24 @@ class IAsset : virtual public core::IReferenceCounted
 
         virtual core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const = 0;
 
-		//! restore() is not recursive!
-		virtual bool restore(IAsset* _other)
+		bool restoreFromDummy(IAsset* _other, uint32_t _levelsBelow)
 		{
-			if (!this->isADummyObjectForCache())
+			if (!canBeRestoredFrom(_other))
+				return false;
+
+			restoreFromDummy_impl(_other, _levelsBelow);
```

Comment:
you should probably split this into `restoreFromDummy_this` and `restoreFromDummy_recurse`, to solve comment
https://github.com/buildaworldnet/IrrlichtBAW/pull/549/commits/750f5f9dfa92c3d1e05482fca38a5723768a4682#r507655849

you can only restore `this` from `_other` if:
1. `_other` matches `this`
2. `this` is dummy
3. `this` is mutable
4. `_other` is not a dummy
5. `_other` is mutable

You can only recurse if **`_other` "matches" `_this`, no other conditions necessary**